src/goto-symex/symex_clean_expr.cpp

```diff
@@ -137,10 +60,57 @@ void goto_symext::process_array_expr(exprt &expr)
     const index_exprt &index_expr=to_index_expr(ssa.get_original_expr());
     exprt tmp=index_expr.array();
     expr.swap(tmp);
+
+    process_array_expr(expr);
+  }
+  else if(expr.id() != ID_symbol)
+  {
+    object_descriptor_exprt ode;
+    ode.build(expr, ns);
+    do_simplify(ode.offset());
+
+    expr = ode.root_object();
+
+    if(!ode.offset().is_zero())
+    {
+      if(expr.type().id() != ID_array)
+      {
+        exprt array_size = size_of_expr(expr.type(), ns);
+        do_simplify(array_size);
+        expr =
+          byte_extract_exprt(
+            byte_extract_id(),
+            expr,
+            from_integer(0, index_type()),
+            array_typet(char_type(), array_size));
+      }
+
+      // given an array type T[S] and a byte offset o, build a new type
+      // T[S-(o/sizeof(T))]
```

Comment:
I may have been particularly slow reading this comment (apologies if so) - but it wasn't immediately obvious to me that this it of type conversion (particularly the `o/sizeof(T)`) is to do the conversion from byte offset -> index. Maybe it's worth expanding the comment very slightly?